WR #463673: Automated Monitoring of Inactive Accounts

waleedhassan5 opened the pull request:
Added All the basic functionalities and php unit tests for the plugin

waleedhassan5 committed:
WR #463673: Automated Monitoring of Inactive Accounts

Added All the basic functionalities and php unit tests for the
plugin

michaelkotlyar commented:
Having a look now, in the meantime here's a few PHPCS warnings

```
FILE: ./admin/tool/inactiveusersgc/tests/scenarios_test.php
------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 19 ERRORS AND 18 WARNINGS AFFECTING 28 LINES
------------------------------------------------------------------------------------------------------------------------------------------------------------
  33 | WARNING | [ ] PHPUnit class "scenarios_test" does not have any namespace. It is recommended to add it to the "tool_inactiveusersgc" namespace, using
     |         |     more levels if needed, in order to match the code being tested
 183 | WARNING | [ ] Test method test_no_action_before_first_window() is missing any coverage information, own or at class level
 201 | WARNING | [ ] Test method test_never_logged_in_uses_timecreated() is missing any coverage information, own or at class level
 220 | WARNING | [ ] Test method test_first_warning_repeats_as_configured() is missing any coverage information, own or at class level
 258 | WARNING | [ ] Test method test_progression_sends_on_stage_increase() is missing any coverage information, own or at class level
 262 | ERROR   | [x] Expected 1 space after comma in argument list; 3 found
 262 | ERROR   | [x] Expected 1 space between the comma and "60". Found: 3 spaces
 262 | ERROR   | [x] Expected 1 space after comma in argument list; 2 found
 262 | ERROR   | [x] Expected 1 space between the comma and "'tool_inactiveusersgc'". Found: 2 spaces
 263 | ERROR   | [x] Expected 1 space after comma in argument list; 2 found
 263 | ERROR   | [x] Expected 1 space between the comma and "300". Found: 2 spaces
 264 | ERROR   | [x] Expected 1 space after comma in argument list; 3 found
 264 | ERROR   | [x] Expected 1 space between the comma and "364". Found: 3 spaces
 265 | ERROR   | [x] Expected 1 space after comma in argument list; 2 found
 265 | ERROR   | [x] Expected 1 space between the comma and "365". Found: 2 spaces
 289 | ERROR   | [x] Expected 1 space after comma in argument list; 2 found
 289 | ERROR   | [x] Expected 1 space between the comma and "360". Found: 2 spaces
 305 | WARNING | [ ] Test method test_final_has_no_repeat() is missing any coverage information, own or at class level
 334 | WARNING | [ ] Test method test_action_suspend() is missing any coverage information, own or at class level
 359 | WARNING | [ ] Test method test_action_delete() is missing any coverage information, own or at class level
 395 | WARNING | [ ] Test method test_tenant_filtering() is missing any coverage information, own or at class level
 423 | WARNING | [ ] Test method test_skips_suspended_and_deleted() is missing any coverage information, own or at class level
 473 | WARNING | [ ] Test method test_observer_clears_tracking_on_login() is missing any coverage information, own or at class level
 514 | WARNING | [ ] Test method test_summary_email_counts() is missing any coverage information, own or at class level
 555 | ERROR   | [x] Expected 1 space after comma in argument list; 4 found
 555 | ERROR   | [x] Expected 1 space between the comma and "$summary". Found: 4 spaces
 565 | WARNING | [ ] Test method test_direct_action_no_extra_email() is missing any coverage information, own or at class level
 621 | WARNING | [ ] Test method test_stage_and_repeats_consistency() is missing any coverage information, own or at class level
 660 | WARNING | [ ] Missing docblock for function dump_emails in testcase
 665 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
 672 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
 677 | WARNING | [ ] Test method test_debug_email_output() is missing any coverage information, own or at class level
 697 | ERROR   | [ ] Variable "userMsgs" must be all lower-case
 697 | ERROR   | [ ] Variable "summaryMsgs" must be all lower-case
 698 | ERROR   | [ ] Variable "userMsgs" must be all lower-case
 699 | ERROR   | [ ] Variable "summaryMsgs" must be all lower-case
 704 | ERROR   | [x] The closing brace for the class must go on the next line after the body
------------------------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 15 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------------------------------------------------------------------------------------


FILE: ./admin/tool/inactiveusersgc/tests/processor_test.php
------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
------------------------------------------------------------------------------------------------------------------------------------------------------------
 39 | WARNING | PHPUnit class "processor_test" does not have any namespace. It is recommended to add it to the "tool_inactiveusersgc" namespace, using more
    |         | levels if needed, in order to match the code being tested
------------------------------------------------------------------------------------------------------------------------------------------------------------


FILE: ./admin/tool/inactiveusersgc/classes/local/processor.php
----------------------------------------------------------------------------------------------------------
FOUND 24 ERRORS AFFECTING 11 LINES
----------------------------------------------------------------------------------------------------------
  75 | ERROR | [x] Expected at least 1 space before "."; 0 found
  75 | ERROR | [x] Expected at least 1 space after "."; 0 found
  75 | ERROR | [x] Expected at least 1 space before "."; 0 found
  75 | ERROR | [x] Expected at least 1 space after "."; 0 found
  75 | ERROR | [x] Expected at least 1 space before "."; 0 found
  75 | ERROR | [x] Expected at least 1 space after "."; 0 found
  96 | ERROR | [x] Long list syntax is not allowed
 196 | ERROR | [x] Opening parenthesis of a multi-line function call must be the last content on the line
 197 | ERROR | [x] Multi-line function call not indented correctly; expected 12 spaces but found 16
 197 | ERROR | [x] Only one argument is allowed per line in a multi-line function call
 197 | ERROR | [x] Closing parenthesis of a multi-line function call must be on a line by itself
 198 | ERROR | [x] Opening parenthesis of a multi-line function call must be the last content on the line
 199 | ERROR | [x] Multi-line function call not indented correctly; expected 12 spaces but found 16
 199 | ERROR | [x] Closing parenthesis of a multi-line function call must be on a line by itself
 200 | ERROR | [x] Opening parenthesis of a multi-line function call must be the last content on the line
 201 | ERROR | [x] Multi-line function call not indented correctly; expected 12 spaces but found 16
 201 | ERROR | [x] Only one argument is allowed per line in a multi-line function call
 201 | ERROR | [x] Closing parenthesis of a multi-line function call must be on a line by itself
 203 | ERROR | [x] Opening parenthesis of a multi-line function call must be the last content on the line
 204 | ERROR | [x] Multi-line function call not indented correctly; expected 12 spaces but found 16
 204 | ERROR | [x] Only one argument is allowed per line in a multi-line function call
 204 | ERROR | [x] Closing parenthesis of a multi-line function call must be on a line by itself
 441 | ERROR | [x] Expected 1 space after comma in argument list; 4 found
 441 | ERROR | [x] Expected 1 space between the comma and "'tool_inactiveusersgc'". Found: 4 spaces
----------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 24 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------------------------------------------


FILE: ./admin/tool/inactiveusersgc/lang/en/tool_inactiveusersgc.php
---------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
---------------------------------------------------------------------------------------------------------------------------------------------
 63 | WARNING | [x] The string key "settings:tenantfilterheading" is not in the correct order, it should be before "task:process"
 64 | WARNING | [x] The string key "settings:noprofilefield" is not in the correct order, it should be before "settings:tenantfilterheading"
---------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
---------------------------------------------------------------------------------------------------------------------------------------------

Time: 425ms; Memory: 24MB
```

michaelkotlyar commented on tests/scenarios_test.php:
```diff
@@ -0,0 +1,704 @@
+        $this->assertSame(1, (int)$rec3->stage, 'Stage remains first while still in first window');
+    }
+
+    protected function dump_emails(array $messages): void {
+        foreach ($messages as $m) {
+            $to = $m->to ?? ($m->toemail ?? '');
+            $subject = $m->subject ?? '';
+            $body = $m->body ?? '';
+            // Common alternates across Moodle versions:
+            $alt = $m->altbody ?? ($m->bodyplain ?? $m->fullmessage ?? $m->fullmessagehtml ?? '');
+            fwrite(STDOUT, "\n==== EMAIL DEBUG ====\n");
+            fwrite(STDOUT, "To: {$to}\n");
+            fwrite(STDOUT, "Subject: {$subject}\n");
+            fwrite(STDOUT, "Body:\n" . ($body !== '' ? $body : $alt) . "\n");
+            // Optional: show all properties for discovery
+            // fwrite(STDOUT, print_r(get_object_vars($m), true));
+        }
+    }
+
```
Is there are reason for this? It meddles with the phpunit output..

waleedhassan5 replied:
ah my bad added that to test email outputs these 2 are extra i ll remove

michaelkotlyar commented on classes/observer.php:
```diff
@@ -0,0 +1,70 @@
+     * Accept base so it works both from Moodle's dispatcher and in direct calls from tests.
+     *
+     * @param base $event
+     * @return void
```
Now this is totally optional, and I don't think it shows up in PHPCS but unless your function explicitly does this: `return;` you don't add `@return void`. I see a few other instances of this in the code. Up to you if you want to change them - https://moodledev.io/general/development/policies/codingstyle#types

michaelkotlyar commented on lang/en/tool_inactiveusersgc.php:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @package    tool_inactiveusersgc
+ * @category   string
+ * @copyright  2025 Waleed ul Hassan
```
where's ya email! (also missing in settings.php)

michaelkotlyar commented on classes/local/processor.php:
```diff
@@ -0,0 +1,445 @@
+            $tenantfilterwhere = '';
+        }
+
+        $sql = "
```
No biggy but the indentation is not following moodle's guidelines https://moodledev.io/general/development/policies/codingstyle/sql#indentation

michaelkotlyar commented on classes/local/processor.php:
```diff
@@ -0,0 +1,445 @@
+            $params['pfshort'] = 'primary_membership_code';
+            $tenantfilterwhere = " AND uid.data $insql ";
+        } else {
+            $tenantfilterwhere = '';
```
You could move `$tenantfilterwhere = '';` before the if statement like you've placed `$tenantfilter = '';`, removing the need for the else statement

michaelkotlyar commented on classes/local/processor.php:
```diff
@@ -0,0 +1,445 @@
+     * @throws dml_exception
+     */
+    protected function find_candidates(): array {
+        global $CFG;
```
Don't need this, right?

michaelkotlyar commented on classes/local/processor.php:
```diff
@@ -0,0 +1,445 @@
+        } else if ($daysinactive >= $cfg->firstdays) {
+            return 1;
+        }
+        return 0;
```
Using `match` might be a little tidier https://www.php.net/manual/en/control-structures.match.php (example 3)

waleedhassan5 committed:
WR463673-Automated-Monitoring-of-Inactive-Accounts

Peer review changes

michaelkotlyar commented on lang/en/tool_inactiveusersgc.php:
```diff
@@ -0,0 +1,64 @@
+ * @license    https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['cap:manage'] = 'Manage inactive users tool settings';
```
This should be `inactiveusersgc:manage`

waleedhassan5 committed:
WR463673-Automated-Monitoring-of-Inactive-Accounts

Peer review feed back changes for some phpcs issues and
coding standards changes
